DataFormats/Provenance/interface/CompactEventAuxiliaryVector.h

```diff
@@ -0,0 +1,188 @@
+    class CompactEventAuxiliaryExtra {
+    public:
+      CompactEventAuxiliaryExtra(bool isReal, ExperimentType eType, int storeNum)
+          : processHistoryID_(), isRealData_(isReal), experimentType_(eType), storeNumber_(storeNum) {}
+      CompactEventAuxiliaryExtra(const EventAuxiliary& ea)
+          : processHistoryID_(ea.processHistoryID()),
+            isRealData_(ea.isRealData()),
+            experimentType_(ea.experimentType()),
+            storeNumber_(ea.storeNumber()) {}
+
+      bool operator==(const CompactEventAuxiliaryExtra& extra) const {
+        return processHistoryID_ == extra.processHistoryID_ && isRealData_ == extra.isRealData_ &&
+               experimentType_ == extra.experimentType_ && storeNumber_ == extra.storeNumber_;
+      }
+      void write(std::ostream& os) const;
+
+      // Process history ID of the full process history (not the reduced process history)
+      ProcessHistoryID processHistoryID_;
+      // Is this real data (i.e. not simulated)
+      bool isRealData_;
+      // Something descriptive of the source of the data
+      ExperimentType experimentType_;
+      //  The LHC store number
+      int storeNumber_;
+    };
+
+    struct ExtraHash {
+      std::size_t operator()(CompactEventAuxiliaryExtra const& extra) const noexcept {
+        return details::hash_value(
+            extra.processHistoryID_.compactForm(), extra.isRealData_, extra.experimentType_, extra.storeNumber_);
+      }
+    };
+
+    using GUIDmemo = std::unordered_set<std::string>;
+    using extraMemo = std::unordered_set<CompactEventAuxiliaryExtra, ExtraHash>;
+
+    class CompactEventAuxiliary {
+    public:
+      CompactEventAuxiliary(EventID const& theId,
+                            std::string const& processGUID,
+                            Timestamp const& theTime,
+                            int bunchXing,
+                            int orbitNum,
+                            CompactEventAuxiliaryExtra const& extra,
+                            GUIDmemo& guidmemo,
+                            extraMemo& extramemo)
+          : id_(theId),
+            processGUID_(memoize(processGUID, guidmemo)),
+            time_(theTime),
+            bunchCrossing_(bunchXing),
+            orbitNumber_(orbitNum),
+            extra_(memoize(extra, extramemo)) {}
+      CompactEventAuxiliary(const EventAuxiliary& ea, GUIDmemo& guidmemo, extraMemo& extramemo)
+          : id_(ea.id()),
+            processGUID_(memoize(ea.processGUID(), guidmemo)),
+            time_(ea.time()),
+            bunchCrossing_(ea.bunchCrossing()),
+            orbitNumber_(ea.orbitNumber()),
+            extra_(memoize(CompactEventAuxiliaryExtra(ea), extramemo)) {}
+
+      void write(std::ostream& os) const;
+
+      ProcessHistoryID const& processHistoryID() const { return extra_.processHistoryID_; }
+      EventID const& id() const { return id_; }
+      std::string const& processGUID() const { return processGUID_; }
```

Comment:
I'm surprised processGUID isn't something that is the same for many, many events

Reply:
Ah, you are holding it by reference.

Reply:
Real data can have a surprisingly large number of unique processGUIDs, simulated data it is generally the same for many events